EGM-certified electron IDs for 2016 data: cut-based and MVA

ikrav opened the pull request:
This PR contains EGM certified cut-based and MVA electron IDs tuned on 80X samples. They are intended to be used on 2016 data. The electron MVA contains the HZZ MVA used for ICHEP16 and the presently recommended general purpose MVA tuned for better performance given features of a more complete 2016 dataset.

ikrav committed:
New HLT-safe electron cut-based selection in VID, incl. python and a few new cut classes

ikrav committed:
Bug fixes, added debug statements

ikrav committed:
A try to fix segv

ikrav committed:
Fixed AOD vs miniAOD detection

ikrav committed:
Corrected cut settings

ikrav committed:
New cut-based electron ID tuned on 80X

ikrav committed:
Fixed a text comment

ikrav committed:
Fixed a typo in cut values

ikrav committed:
Merged egm_id_80X_v1 from repository ikrav

ikrav committed:
Added 80X electron MVA implementation. Restructured location of MVA class files (including old ones) to make things work better for FWLite.

Werbellin committed:
Added general purpose ID

Werbellin committed:
Remove preliminary HZZ setup

Werbellin committed:
Update documentation, bugfixes

ikrav committed:
Merged egm_id_80X_v2 from repository ikrav

ikrav committed:
Adjused MD5 sums of the electron MVA IDs

ikrav committed:
A few changes to make new electron MVA work in FWLite

ikrav committed:
Removed a header from an MVA base class that appears unnecessary and was preventing the release from building in 81X+

cmsbuild commented:
A new Pull Request was created by  @ikrav  for CMSSW_9_0_X.

It involves the following packages:

RecoEgamma/EgammaTools
RecoEgamma/ElectronIdentification


@cmsbuild, @cvuosalo, @slava77, @davidlange6 can you please review it and eventually sign? Thanks.
@Sam-Harper, @rafaellopesdesa, @lgray this is something you requested to watch as well.
@slava77, @smuzaffar you are the release manager for this.

cms-bot commands are listed here cms-sw/cmssw#13028

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e775e89f1115adb84d07fd2b31861db30c9be2ee
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16839/16734/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-16839/16734/summary.html

The workflows 1003.0, 1001.0, 1000.0, 140.53, 136.731, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

slava77 commented on RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h:
```diff
@@ -0,0 +1,155 @@
+
+#include <vector>
+#include <string>
+#include <TROOT.h>
```
is TROOT include really needed?

cvuosalo commented:
This PR needs to be re-based.

ikrav commented:
TROOT.h is included in all past MVA headers of this type for electron/photon IDs. 
It was included from the start because TMVA tutorials always include it. I do not know if it
is totally necessary to include it, I guess I can check and see if it works without it. 

> On Dec 5, 2016, at 4:39 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> @slava77 commented on this pull request.
> 
> In RecoEgamma/ElectronIdentification/interface/ElectronMVAEstimatorRun2Spring16GeneralPurpose.h:
> 
> > +
> +#include "RecoEgamma/EgammaTools/interface/AnyMVAEstimatorRun2Base.h"
> +
> +#include "DataFormats/EgammaCandidates/interface/GsfElectron.h"
> +
> +#include "DataFormats/BeamSpot/interface/BeamSpot.h"
> +
> +#include "DataFormats/EgammaCandidates/interface/ConversionFwd.h"
> +#include "DataFormats/EgammaCandidates/interface/Conversion.h"
> +#include "RecoEgamma/EgammaTools/interface/ConversionTools.h"
> +
> +#include "CondFormats/EgammaObjects/interface/GBRForest.h"
> +
> +#include <vector>
> +#include <string>
> +#include <TROOT.h>
> 
> is TROOT include really needed?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ikrav commented:
Carl, does this mean I need to do something? If yes, can you give me
guidance?

> On Dec 5, 2016, at 5:21 PM, Carl Vuosalo <notifications@github.com> wrote:
> 
> This PR needs to be re-based.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
@ikrav 
from tomorrow you should be able to test and submit a pull request in 80X.
The data files should make it to the new IB (probably CMSSW_8_0_X-12-07-2300)

For rebase instructions you may want to check 
https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe

Your proposed code change is currently in conflict with  RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.h  file in 90X release (as noted at the bottom of this PR thread web page https://github.com/cms-sw/cmssw/pull/16839 )

ikrav commented:
What should I rebase? The PR I submitted to 90X? Or the PR I am about
to submit to 80X? All of them? Is this a rule, do I need to rebase anything to
the release to which I am submitting the PR before I submit the PR?
-thanks

> On Dec 7, 2016, at 12:36 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> @ikrav
> from tomorrow you should be able to test and submit a pull request in 80X.
> The data files should make it to the new IB (probably CMSSW_8_0_X-12-07-2300)
> 
> For rebase instructions you may want to check
> https://twiki.cern.ch/twiki/bin/viewauth/CMS/SWGuideHATS2016Git#CMSSW_rebasing_recipe
> 
> Your proposed code change is currently in conflict with RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.h file in 90X release (as noted at the bottom of this PR thread web page #16839 )
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
On 12/7/16 12:12 PM, ikrav wrote:
> What should I rebase? The PR I submitted to 90X? Or the PR I am about
> to submit to 80X? All of them? Is this a rule, do I need to rebase
> anything to
> the release to which I am submitting the PR before I submit the PR?
> -thanks


You need to rebase the PR that exists and is said to have a merge
conflict (this 90X PR).

Ideally, you need to prepare the PR topic branch in the same release
that you submit it to.

ikrav commented:
Ok, I will rebase the 90X PR. Sorry for many questions, but what is the merge conflict
you see? I merged my private branch into 90X IB most recent release when I was submitting,
and it merged fine and built fine.



> On Dec 7, 2016, at 2:21 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 12/7/16 12:12 PM, ikrav wrote:
> > What should I rebase? The PR I submitted to 90X? Or the PR I am about
> > to submit to 80X? All of them? Is this a rule, do I need to rebase
> > anything to
> > the release to which I am submitting the PR before I submit the PR?
> > -thanks
> 
> 
> You need to rebase the PR that exists and is said to have a merge
> conflict (this 90X PR).
> 
> Ideally, you need to prepare the PR topic branch in the same release
> that you submit it to.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

slava77 commented:
On 12/7/16 12:25 PM, ikrav wrote:
> Ok, I will rebase the 90X PR. Sorry for many questions, but what is the
> merge conflict
> you see? I merged my private branch into 90X IB most recent release when
> I was submitting,
> and it merged fine and built fine.

I did not look which lines actually conflict.
There is a message from automated checking
 "This branch has conflicts that must be resolved"
I hope you can see it.

cvuosalo commented on RecoEgamma/ElectronIdentification/src/ElectronMVAEstimatorRun2Spring16GeneralPurpose.cc:
```diff
@@ -0,0 +1,460 @@
+
+ElectronMVAEstimatorRun2Spring16GeneralPurpose::ElectronMVAEstimatorRun2Spring16GeneralPurpose(const edm::ParameterSet& conf):
+  AnyMVAEstimatorRun2Base(conf),
+  _tag(conf.getParameter<std::string>("mvaTag")),
```
Variable names starting with "_" violate the CMS coding rules. They need to be changed.

ikrav commented:
> On Dec 7, 2016, at 2:37 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> I did not look which lines actually conflict.
> There is a message from automated checking
> "This branch has conflicts that must be resolved"
> I hope you can see it.

I do see now that:

This branch has conflicts that must be resolved
Only those with write access to this repository can merge pull requests.
Conflicting files
RecoEgamma/ElectronIdentification/plugins/ElectronMVAEstimatorRun2Phys14NonTrig.h

But how can I learned more about it? I just checked out a fresh CMSSW_9_0_X_2016-12-07-0000.
I merged into it my branch ikrav:egm_id_80X_v2 which I used for this PR. I see no merging
conflicts. The GitHub page of the PR tells me no information beyond the name of the
file above. How can I find what and where is wrong?

cvuosalo commented on RecoEgamma/ElectronIdentification/python/ElectronMVAValueMapProducer_cfi.py:
```diff
@@ -15,6 +15,13 @@
 from RecoEgamma.ElectronIdentification.Identification.mvaElectronID_Spring15_25ns_Trig_V1_cff import *
 mvaConfigsForEleProducer.append( mvaEleID_Spring15_25ns_Trig_V1_producer_config )
 
+from RecoEgamma.ElectronIdentification.Identification.mvaElectronID_Spring16_HZZ_V1_cff import *
```
Please don't use `import *`. Please list the modules you need to import. There are many instances of `import *` in this PR that should be fixed.

davidlange6 commented:
i don’t see a big need to avoid import * - standard python practice. Is this cff particularly big?

> On Dec 8, 2016, at 1:38 PM, Carl Vuosalo <notifications@github.com> wrote:
> 
> @cvuosalo commented on this pull request.
> 
> In RecoEgamma/ElectronIdentification/python/ElectronMVAValueMapProducer_cfi.py:
> 
> > @@ -15,6 +15,13 @@
>  from RecoEgamma.ElectronIdentification.Identification.mvaElectronID_Spring15_25ns_Trig_V1_cff import *
>  mvaConfigsForEleProducer.append( mvaEleID_Spring15_25ns_Trig_V1_producer_config )
>  
> +from RecoEgamma.ElectronIdentification.Identification.mvaElectronID_Spring16_HZZ_V1_cff import *
> 
> Please don't use import *. Please list the modules you need to import. There are many instances of import * in this PR that should be fixed.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ikrav commented:
Ok, I followed instructions for rebasing and created a new branch in my-cmssw
based on a recent 90X release. Now the question is what to do with it. I am reluctant
to update the original branch that was used to submit the PR (based on 8_0_10) 
with the rebased version because it is already publicized in recipes of electron IDs. 
Any advice?


> On Dec 7, 2016, at 2:21 PM, Slava Krutelyov <notifications@github.com> wrote:
> 
> On 12/7/16 12:12 PM, ikrav wrote:
> > What should I rebase? The PR I submitted to 90X? Or the PR I am about
> > to submit to 80X? All of them? Is this a rule, do I need to rebase
> > anything to
> > the release to which I am submitting the PR before I submit the PR?
> > -thanks
> 
> 
> You need to rebase the PR that exists and is said to have a merge
> conflict (this 90X PR).
> 
> Ideally, you need to prepare the PR topic branch in the same release
> that you submit it to.
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cvuosalo commented:
Reco would prefer to reduce the overhead of importing and constructing large numbers of unused modules. If the `import *` only imports modules that are truly needed, then it might be OK, but later additions to the imported file could then be imported unnecessarily.

cvuosalo commented:
A test of workflow 25202.0_TTbar_13 with no DQM or Validation with 150 events against baseline CMSSW_9_0_0_pre1 shows no significant differences and no increase in memory usage. CPU time increases very marginally:
```
  delta/mean delta/orJob     original                   new       module name
  ---------- ------------     --------                  ----       ------------
   +0.478014      +0.01%         0.62 ms/ev ->         1.00 ms/ev electronMVAValueMapProducer
  ---------- ------------     --------                  ----       ------------
```

ikrav commented:
I replaced several import * by explicit lists (only the new ones, though. I have not
changed those import * that are already in CMSSW). Additionally, I removed TROOT.h
includes, another request. Finally, names with leading underscores are renamed. 
All of this is in my new rebased to 90X branch named differently from the branch that
started this PR. Now my question is how to proceed (same question was also in another
comment from me from a few hours ago).


> On Dec 8, 2016, at 1:51 PM, Carl Vuosalo <notifications@github.com> wrote:
> 
> Reco would prefer to reduce the overhead of importing and constructing large numbers of unused modules. If the import * only imports modules that are truly needed, then it might be OK, but later additions to the imported file could then be imported unnecessarily.

cvuosalo commented:
@ikrav: If you're still having problems with your new branch, you could just create a new PR and then close this one.

cvuosalo commented:
@ikrav: If you want this PR in 900pre2, it needs to be fixed today (Friday).

ikrav commented:
Carl, I do not have problems with my new branch. Rather, I do not know what to do 
with it. I have the branch that I used to create the PR. I do not want to change
that branch if possible because it is already being used by people in recipes.
I created a new branch with changes you and Slava requested. Now my question
is whether somehow this PR can use the new branch, or whether I have to cancel
this one and start new one, or whether a new PR can be started from this PR if such
a thing is possible, etc. I do not have experience, so I do not know what my steps are,
and it seems that there are not too many directly involved experts in my time zone.


> On Dec 9, 2016, at 8:30 AM, Carl Vuosalo <notifications@github.com> wrote:
> 
> @ikrav: If you're still having problems with your new branch, you could just create a new PR and then close this one.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

cvuosalo commented:
@ikrav: Creating a new PR would seem the simplest approach. I'm not understanding why that hasn't been done yet.

ikrav commented:
Carl, this has not been done because I was waiting for explicit guidance from you
or Slava. I will create a new PR today. 

> On Dec 11, 2016, at 4:41 PM, Carl Vuosalo <notifications@github.com> wrote:
> 
> @ikrav: Creating a new PR would seem the simplest approach. I'm not understanding why that hasn't been done yet.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

ikrav commented:
This PR is closed as I just submitted a new PR from a rebased to 90X branch. The new PR number is #16979.
